test/gtest/parser_unroll.cpp

```diff
@@ -0,0 +1,297 @@
+                std::cout << toplevel->getMain() << std::endl;
+            },
+            new P4::MidEndLast()
+        });
+        if (options.listMidendPasses) {
+            listPasses(*outStream, "\n");
+            *outStream << std::endl;
+            return;
+        }
+        if (options.excludeMidendPasses) {
+            removePasses(options.passesToExcludeMidend);
+        }
+        toplevel = evaluator->getToplevelBlock();
+    }
+    IR::ToplevelBlock* process(const IR::P4Program *&program) {
+        program = program->apply(*this);
+        return toplevel; }
+};
+
+/// Relative path to the examples
+const char *relPath = "./testdata/p4_16_samples/parserUnroll/";
+
+// #define PARSER_UNROLL_TIME_CHECKING
+
+#ifdef PARSER_UNROLL_TIME_CHECKING
+    #include <chrono>
+#endif
+
+/// Rewrites parser
+std::pair<const IR::P4Parser*, const IR::P4Parser*> rewriteParser(const IR::P4Program* program,
+                                                                  CompilerOptions& options) {
+    P4::FrontEnd frontend;
+    program = frontend.run(options, program);
+    P4::ReferenceMap    refMap;
+    P4::TypeMap         typeMap;
+
+    const IR::P4Program* res = nullptr;
+#ifdef PARSER_UNROLL_TIME_CHECKING
+    using std::chrono::high_resolution_clock;
+    using std::chrono::duration_cast;
+    using std::chrono::duration;
+    using std::chrono::milliseconds;
+    auto t1 = high_resolution_clock::now();
+#endif
+    ParsersUnroll unroller(true, &refMap, &typeMap);
+    res  = program->apply(unroller);
+    std::cout << res << std::endl;
+#ifdef PARSER_UNROLL_TIME_CHECKING
+    auto t2 = high_resolution_clock::now();
+    auto msInt = duration_cast<milliseconds>(t2 - t1);
+    std::cout << msInt.count() << std::endl;
+#endif
+    const auto newDeclVector = res->getDeclsByName("MyParser")->toVector();
+    const auto* newDecl = (*newDeclVector)[0];
+    const auto oldDeclVector = program->getDeclsByName("MyParser")->toVector();
+    const auto* oldDecl = (*oldDeclVector)[0];
+    return std::make_pair(oldDecl->to<IR::P4Parser>(), newDecl->to<IR::P4Parser>());
+}
+
+/// Loads example from a file
+const IR::P4Program* load_model(const char* curFile, CompilerOptions& options) {
+    options.langVersion = CompilerOptions::FrontendVersion::P4_16;
+    options.compilerVersion = P4TEST_VERSION_STRING;
+    options.file = relPath;
+    options.file += curFile;
+    return P4::parseP4File(options);
+}
+
+std::pair<const IR::P4Parser*, const IR::P4Parser*> loadExample(const char *file) {
+     AutoCompileContext autoP4TestContext(new P4TestContext);
+    auto& options = P4CContext::get().options();
+    const char* argv = "./gtestp4c";
+    options.process(1, (char* const*)&argv);
+    const IR::P4Program* program = load_model(file, options);
+    if (!program)
+        return std::make_pair(nullptr, nullptr);
+    return rewriteParser(program, options);
+}
+
+TEST_F(P4CParserUnroll, example1) {
+    auto parsers = loadExample("example1.p4");
```

Comment:
Since the file names are explicit here I don't think it's difficult to keep them in the same folder with the other tests.